mtp_api/apps/transaction/api/cashbook/views.py

```diff
@@ -50,117 +53,131 @@ def get_prison_set_for_user(self):
-            slice_pks = pending.values_list('pk', flat=True)[:slice_size]
 
-            queryset = self.get_queryset(filter_by_user=False).filter(pk__in=slice_pks)
-            for t in queryset:
-                t.take(by_user=request.user)
+class LockTransactions(TransactionViewMixin, APIView):
+    action = 'lock'
 
-            return HttpResponseRedirect(
-                reverse('cashbook:transaction-prison-user-list', kwargs=kwargs),
-                status=status.HTTP_303_SEE_OTHER
+    permission_classes = (
+        IsAuthenticated, CashbookClientIDPermissions,
+        TransactionPermissions
+    )
+
+    def post(self, request, format=None):
+        with transaction.atomic():
+            locked_count = self.get_queryset().locked().filter(owner=self.request.user).count()
+            if locked_count < LOCK_LIMIT:
+                slice_size = LOCK_LIMIT-locked_count
+                to_lock = self.get_queryset().available().select_for_update()
+                slice_pks = to_lock.values_list('pk', flat=True)[:slice_size]
+
+                queryset = self.get_queryset().filter(pk__in=slice_pks)
+                for t in queryset:
+                    t.lock(by_user=request.user)
+
+            redirect_url = '{url}?user={user}&status={status}'.format(
+                url=reverse('cashbook:transaction-list'),
+                user=request.user.pk,
+                status=TRANSACTION_STATUS.LOCKED
             )
+            return HttpResponseRedirect(redirect_url, status=drf_status.HTTP_303_SEE_OTHER)
 
-    def get_slice_limit(self, request):
-        slice_size = int(request.query_params.get('count', DEFAULT_SLICE_SIZE))
-        available_to_take = max(0, TAKE_LIMIT - self.queryset.pending().filter(owner=request.user).count())
-        if available_to_take < slice_size:
-            raise exceptions.ParseError(detail='Can\'t take more than %s transactions.' % TAKE_LIMIT)
 
-        return slice_size
+class UnlockTransactions(TransactionViewMixin, APIView):
+    action = 'unlock'
 
-    def release(self, request, *args, **kwargs):
+    permission_classes = (
+        IsAuthenticated, CashbookClientIDPermissions,
+        TransactionPermissions
+    )
 
+    def post(self, request, format=None):
         transaction_ids = request.data.get('transaction_ids', [])
         with transaction.atomic():
-            to_update = self.get_queryset().pending().filter(pk__in=transaction_ids).select_for_update()
+            to_update = self.get_queryset().locked().filter(pk__in=transaction_ids).select_for_update()
             if len(to_update) != len(transaction_ids):
                 return Response(
-                    data={'transaction_ids': ['Some transactions could not be released.']},
-                    status=status.HTTP_400_BAD_REQUEST
+                    data={'transaction_ids': ['Some transactions cannot be unlocked.']},
+                    status=drf_status.HTTP_409_CONFLICT
```

Comment:
Can we list the transaction IDs that couldn't be unlocked? Maybe have the current message as `error`, then `failed_transaction_ids` array?